File: ring-android/libringclient/src/main/java/net/jami/conversation/ConversationPresenter.java
/*
 *  Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 *  Author: Hadrien De Sousa <hadrien.desousa@savoirfairelinux.com>
 *  Author: Adrien Béraud <adrien.beraud@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package net.jami.conversation;

import net.jami.daemon.Blob;
import net.jami.facades.ConversationFacade;
import net.jami.model.Account;
import net.jami.model.Call;
import net.jami.model.Conference;
import net.jami.model.Contact;
import net.jami.model.Conversation;
import net.jami.model.DataTransfer;
import net.jami.model.Error;
import net.jami.model.Interaction;
import net.jami.model.TrustRequest;
import net.jami.model.Uri;
import net.jami.mvp.RootPresenter;
import net.jami.services.AccountService;
import net.jami.services.ContactService;
import net.jami.services.DeviceRuntimeService;
import net.jami.services.HardwareService;
import net.jami.services.PreferencesService;
import net.jami.services.VCardService;
import net.jami.utils.Log;
import net.jami.utils.StringUtils;
import net.jami.utils.Tuple;
import net.jami.utils.VCardUtils;

import java.io.File;

import javax.inject.Inject;
import javax.inject.Named;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;
import io.reactivex.rxjava3.subjects.BehaviorSubject;
import io.reactivex.rxjava3.subjects.Subject;

public class ConversationPresenter extends RootPresenter<ConversationView> {

    private static final String TAG = ConversationPresenter.class.getSimpleName();
    private final ContactService mContactService;
    private final AccountService mAccountService;
    private final HardwareService mHardwareService;
    private final ConversationFacade mConversationFacade;
    private final VCardService mVCardService;
    private final DeviceRuntimeService mDeviceRuntimeService;
    private final PreferencesService mPreferencesService;

    private Conversation mConversation;
    private Uri mConversationUri;

    private CompositeDisposable mConversationDisposable;
    private final CompositeDisposable mVisibilityDisposable = new CompositeDisposable();

    @Inject
    @Named("UiScheduler")
    protected Scheduler mUiScheduler;

    private final Subject<Conversation> mConversationSubject = BehaviorSubject.create();

    @Inject
    public ConversationPresenter(ContactService contactService,
                                 AccountService accountService,
                                 HardwareService hardwareService,
                                 ConversationFacade conversationFacade,
                                 VCardService vCardService,
                                 DeviceRuntimeService deviceRuntimeService,
                                 PreferencesService preferencesService) {
        mContactService = contactService;
        mAccountService = accountService;
        mHardwareService = hardwareService;
        mConversationFacade = conversationFacade;
        mVCardService = vCardService;
        mDeviceRuntimeService = deviceRuntimeService;
        mPreferencesService = preferencesService;
        mCompositeDisposable.add(mVisibilityDisposable);
    }

    public void init(Uri conversationUri, String accountId) {
        Log.w(TAG, "init " + conversationUri + " " + accountId);
        if (conversationUri.equals(mConversationUri))
            return;
        mConversationUri = conversationUri;
        mCompositeDisposable.add(mConversationFacade.getAccountSubject(accountId)
                .flatMap(a -> mConversationFacade.loadConversationHistory(a, conversationUri)
                        .observeOn(mUiScheduler)
                        .doOnSuccess(c -> setConversation(a, c)))
                .observeOn(mUiScheduler)
                .subscribe(c -> {}, e -> {
                    Log.e(TAG, "Error loading conversation", e);
                    getView().goToHome();
                }));
        getView().setReadIndicatorStatus(showReadIndicator());
    }

    @Override
    public void unbindView() {
        super.unbindView();
        mConversation = null;
        mConversationUri = null;
        if (mConversationDisposable != null) {
            mConversationDisposable.dispose();
            mConversationDisposable = null;
        }
    }

    private void setConversation(Account account, final Conversation conversation) {
        Log.w(TAG, "setConversation " + conversation.getAggregateHistory().size());
        if (mConversation == conversation)
            return;
        mConversation = conversation;
        mConversationSubject.onNext(conversation);
        ConversationView view = getView();
        if (view != null)
            initView(account, conversation, view);
    }

    public void pause() {
        mVisibilityDisposable.clear();
        if (mConversation != null) {
            mConversation.setVisible(false);
        }
    }

    public void resume(boolean isBubble) {
        Log.w(TAG, "resume " + mConversationUri);
        mVisibilityDisposable.clear();
        mVisibilityDisposable.add(mConversationSubject
                .subscribe(conversation -> {
                    conversation.setVisible(true);
                    updateOngoingCallView(conversation);
                    mConversationFacade.readMessages(mAccountService.getAccount(conversation.getAccountId()), conversation, !isBubble);
                }, e -> Log.e(TAG, "Error loading conversation", e)));
    }

    private void initContact(final Account account, final Conversation conversation, Conversation.Mode mode, final ConversationView view) {
        if (account.isJami()) {
            Log.w(TAG, "initContact " + conversation.getUri() + " mode:" + mode + " " + conversation.getContacts());
            if (mode == Conversation.Mode.Syncing) {
                view.switchToSyncingView();
            } else if (conversation.isSwarm() || account.isContact(conversation)) {
                //if (conversation.isEnded())
                //    conversation.s
                view.switchToConversationView();
            } else {
                Uri uri = conversation.getUri();
                TrustRequest req = account.getRequest(uri);
                if (req == null) {
                    view.switchToUnknownView(uri.getRawUriString());
                } else {
                    view.switchToIncomingTrustRequestView(req.getDisplayname());
                }
            }
        } else {
            view.switchToConversationView();
        }
        view.displayContact(conversation);
    }

    private void initView(Account account, final Conversation c, final ConversationView view) {
        Log.w(TAG, "initView " + c.getUri() + " " + c.getMode());
        if (mConversationDisposable == null) {
            mConversationDisposable = new CompositeDisposable();
            mCompositeDisposable.add(mConversationDisposable);
        }
        mConversationDisposable.clear();
        view.hideNumberSpinner();

        mConversationDisposable.add(c.getMode()
                .switchMapSingle(mode -> mContactService.getLoadedContact(c.getAccountId(), c.getContacts(), true)
                        .observeOn(mUiScheduler)
                        .doOnSuccess(contact -> initContact(account, c, mode, view)))
                .subscribe());
        mConversationDisposable.add(c.getMode()
                .switchMap(mode -> (mode == Conversation.Mode.Legacy || mode == Conversation.Mode.OneToOne) ?
                        c.getContact().getConversationUri() : Observable.empty())
                .observeOn(mUiScheduler)
                .subscribe(uri -> init(uri, account.getAccountID())));

        mConversationDisposable.add(Observable.combineLatest(
                mHardwareService.getConnectivityState(),
                mAccountService.getObservableAccount(account),
                (isConnected, a) -> isConnected || a.isRegistered())
                .observeOn(mUiScheduler)
                .subscribe(isOk -> {
                    ConversationView v = getView();
                    if (v != null) {
                        if (!isOk)
                            v.displayNetworkErrorPanel();
                        else if(!account.isEnabled()) {
                            v.displayAccountOfflineErrorPanel();
                        }
                        else {
                            v.hideErrorPanel();
                        }
                    }
                }));

        mConversationDisposable.add(c.getSortedHistory()
                .observeOn(mUiScheduler)
                .subscribe(view::refreshView, e -> Log.e(TAG, "Can't update element", e)));
        mConversationDisposable.add(c.getCleared()
                .observeOn(mUiScheduler)
                .subscribe(view::refreshView, e -> Log.e(TAG, "Can't update elements", e)));

        mConversationDisposable.add(c.getContactUpdates()
                .switchMap(contacts -> Observable.merge(mContactService.observeLoadedContact(c.getAccountId(), contacts, true)))
                .observeOn(mUiScheduler)
                .subscribe(contact -> {
                    ConversationView v = getView();
                    if (v != null)
                        v.updateContact(contact);
                }));

        mConversationDisposable.add(c.getUpdatedElements()
                .observeOn(mUiScheduler)
                .subscribe(elementTuple -> {
                    switch(elementTuple.second) {
                        case ADD:
                            view.addElement(elementTuple.first);
                            break;
                        case UPDATE:
                            view.updateElement(elementTuple.first);
                            break;
                        case REMOVE:
                            view.removeElement(elementTuple.first);
                            break;
                    }
                }, e -> Log.e(TAG, "Can't update element", e)));

        if (showTypingIndicator()) {
            mConversationDisposable.add(c.getComposingStatus()
                    .observeOn(mUiScheduler)
                    .subscribe(view::setComposingStatus));
        }
        mConversationDisposable.add(c.getLastDisplayed()
                .observeOn(mUiScheduler)
                .subscribe(view::setLastDisplayed));
        mConversationDisposable.add(c.getCalls()
                .observeOn(mUiScheduler)
                .subscribe(calls -> updateOngoingCallView(mConversation), e -> Log.e(TAG, "Can't update call view", e)));
        mConversationDisposable.add(c.getColor()
                .observeOn(mUiScheduler)
                .subscribe(view::setConversationColor, e -> Log.e(TAG, "Can't update conversation color", e)));
        mConversationDisposable.add(c.getSymbol()
                .observeOn(mUiScheduler)
                .subscribe(view::setConversationSymbol, e -> Log.e(TAG, "Can't update conversation color", e)));

        mConversationDisposable.add(account
                .getLocationUpdates(c.getUri())
                .observeOn(mUiScheduler)
                .subscribe(u -> {
                    Log.e(TAG, "getLocationUpdates: update");
                    getView().showMap(c.getAccountId(), c.getUri().getUri(), false);
                }));
    }

    public void loadMore() {
        mConversationDisposable.add(mAccountService.loadMore(mConversation)
                .subscribe(c -> {}, e-> {}));
    }

    public void openContact() {
        if (mConversation != null)
            getView().goToContactActivity(mConversation.getAccountId(), mConversation.getUri());
    }

    public void sendTextMessage(String message) {
        if (StringUtils.isEmpty(message) || mConversation == null) {
            return;
        }
        Conference conference = mConversation.getCurrentCall();
        if (mConversation.isSwarm() || conference == null || !conference.isOnGoing()) {
            mConversationFacade.sendTextMessage(mConversation, mConversationUri, message).subscribe();
        } else {
            mConversationFacade.sendTextMessage(mConversation, conference, message);
        }
    }

    public void selectFile() {
        getView().openFilePicker();
    }

    public void sendFile(File file) {
        if (mConversation ==  null)
            return;
        mConversationFacade.sendFile(mConversation, mConversationUri, file).subscribe();
    }

    /**
     * Gets the absolute path of the file dataTransfer and sends both the DataTransfer and the
     * found path to the ConversationView in order to start saving the file
     *
     * @param interaction an interaction representing a datat transfer
     */
    public void saveFile(Interaction interaction) {
        DataTransfer transfer = (DataTransfer) interaction;
        String fileAbsolutePath = getDeviceRuntimeService().
                getConversationPath(transfer)
                .getAbsolutePath();
        getView().startSaveFile(transfer, fileAbsolutePath);
    }

    public void shareFile(Interaction interaction) {
        DataTransfer file = (DataTransfer) interaction;
        File path = getDeviceRuntimeService().getConversationPath(file);
        getView().shareFile(path, file.getDisplayName());
    }

    public void openFile(Interaction interaction) {
        DataTransfer file = (DataTransfer) interaction;
        File path = getDeviceRuntimeService().getConversationPath(file);
        getView().openFile(path, file.getDisplayName());
    }

    public void acceptFile(DataTransfer transfer) {
        getView().acceptFile(mConversation.getAccountId(), mConversationUri, transfer);
    }

    public void refuseFile(DataTransfer transfer) {
        getView().refuseFile(mConversation.getAccountId(), mConversationUri, transfer);
    }

    public void deleteConversationItem(Interaction element) {
        mConversationFacade.deleteConversationItem(mConversation, element);
    }

    public void cancelMessage(Interaction message) {
        mConversationFacade.cancelMessage(message);
    }

    private void sendTrustRequest() {
        Contact contact = mConversation.getContact();
        if (contact != null) {
            contact.setStatus(Contact.Status.REQUEST_SENT);
        }
        mVCardService.loadSmallVCardWithDefault(mConversation.getAccountId(), VCardService.MAX_SIZE_REQUEST)
                .subscribeOn(Schedulers.computation())
                .subscribe(vCard -> mAccountService.sendTrustRequest(mConversation, contact.getUri(), Blob.fromString(VCardUtils.vcardToString(vCard))),
                        e -> mAccountService.sendTrustRequest(mConversation, contact.getUri(), null));
    }

    public void clickOnGoingPane() {
        Conference conf = mConversation == null ? null : mConversation.getCurrentCall();
        if (conf != null) {
            getView().goToCallActivity(conf.getId());
        } else {
            getView().displayOnGoingCallPane(false);
        }
    }

    public void goToCall(boolean audioOnly) {
        if (audioOnly && !mHardwareService.hasMicrophone()) {
            getView().displayErrorToast(Error.NO_MICROPHONE);
            return;
        }

        mCompositeDisposable.add(mConversationSubject
                .firstElement()
                .subscribe(conversation -> {
                    ConversationView view = getView();
                    if (view != null) {
                        Conference conf = mConversation.getCurrentCall();
                        if (conf != null
                                && !conf.getParticipants().isEmpty()
                                && conf.getParticipants().get(0).getCallStatus() != Call.CallStatus.INACTIVE
                                && conf.getParticipants().get(0).getCallStatus() != Call.CallStatus.FAILURE) {
                            view.goToCallActivity(conf.getId());
                        } else {
                            view.goToCallActivityWithResult(mConversation.getAccountId(), mConversation.getUri(), mConversation.getContact().getUri(), audioOnly);
                        }
                    }
                }));
    }

    private void updateOngoingCallView(Conversation conversation) {
        Conference conf = conversation == null ? null : conversation.getCurrentCall();
        getView().displayOnGoingCallPane(conf != null && (conf.getState() == Call.CallStatus.CURRENT || conf.getState() == Call.CallStatus.HOLD || conf.getState() == Call.CallStatus.RINGING));
    }

    public void onBlockIncomingContactRequest() {
        mConversationFacade.discardRequest(mConversation.getAccountId(), mConversationUri);
        mAccountService.removeContact(mConversation.getAccountId(), mConversationUri.getHost(), true);

        getView().goToHome();
    }

    public void onRefuseIncomingContactRequest() {
        mConversationFacade.discardRequest(mConversation.getAccountId(), mConversationUri);
        getView().goToHome();
    }

    public void onAcceptIncomingContactRequest() {
        mConversationFacade.acceptRequest(mConversation.getAccountId(), mConversationUri);
        getView().switchToConversationView();
    }

    public void onAddContact() {
        sendTrustRequest();
        getView().switchToConversationView();
    }

    public DeviceRuntimeService getDeviceRuntimeService() {
        return mDeviceRuntimeService;
    }

    public void noSpaceLeft() {
        Log.e(TAG, "configureForFileInfoTextMessage: no space left on device");
        getView().displayErrorToast(Error.NO_SPACE_LEFT);
    }

    public void setConversationColor(int color) {
        mCompositeDisposable.add(mConversationSubject
                .firstElement()
                .subscribe(conversation -> conversation.setColor(color)));
    }
    public void setConversationSymbol(CharSequence symbol) {
        mCompositeDisposable.add(mConversationSubject
                .firstElement()
                .subscribe(conversation -> conversation.setSymbol(symbol)));
    }

    public void cameraPermissionChanged(boolean isGranted) {
        if (isGranted && mHardwareService.isVideoAvailable()) {
            mHardwareService.initVideo()
                    .onErrorComplete()
                    .subscribe();
        }
    }

    public void shareLocation() {
        getView().startShareLocation(mConversation.getAccountId(), mConversationUri.getUri());
    }

    public void showPluginListHandlers() {
        getView().showPluginListHandlers(mConversation.getAccountId(), mConversationUri.getUri());
    }

    public Tuple<String, String> getPath() {
        return new Tuple<>(mConversation.getAccountId(), mConversationUri.getUri());
    }

    public void onComposingChanged(boolean hasMessage) {
        if (mConversation == null || !showTypingIndicator()) {
            return;
        }
        mConversationFacade.setIsComposing(mConversation.getAccountId(), mConversationUri, hasMessage);
    }

    public boolean showTypingIndicator() {
        return mPreferencesService.getSettings().isAllowTypingIndicator();
    }

    private boolean showReadIndicator() {
        return mPreferencesService.getSettings().isAllowReadIndicator();
    }

}

File: ring-android/libjamiclient/src/main/java/net/jami/conversation/ConversationPresenter.kt
/*
 *  Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 *  Author: Hadrien De Sousa <hadrien.desousa@savoirfairelinux.com>
 *  Author: Adrien Béraud <adrien.beraud@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package net.jami.conversation

import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.schedulers.Schedulers
import io.reactivex.rxjava3.subjects.BehaviorSubject
import io.reactivex.rxjava3.subjects.Subject
import net.jami.daemon.Blob
import net.jami.services.ConversationFacade
import net.jami.model.*
import net.jami.model.Account.ComposingStatus
import net.jami.model.Conversation.ElementStatus
import net.jami.mvp.RootPresenter
import net.jami.services.*
import net.jami.utils.Log
import net.jami.utils.StringUtils
import net.jami.utils.VCardUtils
import java.io.File
import javax.inject.Inject
import javax.inject.Named

class ConversationPresenter @Inject constructor(
    private val mContactService: ContactService,
    private val mAccountService: AccountService,
    private val mHardwareService: HardwareService,
    private val mConversationFacade: ConversationFacade,
    private val mVCardService: VCardService,
    val deviceRuntimeService: DeviceRuntimeService,
    private val mPreferencesService: PreferencesService,
    @param:Named("UiScheduler") private var mUiScheduler: Scheduler
) : RootPresenter<ConversationView>() {
    private var mConversation: Conversation? = null
    private var mConversationUri: Uri? = null
    private var mConversationDisposable: CompositeDisposable? = null
    private val mVisibilityDisposable = CompositeDisposable()
    private val mConversationSubject: Subject<Conversation> = BehaviorSubject.create()

    fun init(conversationUri: Uri, accountId: String) {
        Log.w(TAG, "init $conversationUri $accountId")
        if (conversationUri == mConversationUri) return
        mConversationUri = conversationUri
        mCompositeDisposable.add(mConversationFacade.getAccountSubject(accountId)
            .flatMap { a: Account ->
                mConversationFacade.loadConversationHistory(a, conversationUri)
                    .observeOn(mUiScheduler)
                    .doOnSuccess { c: Conversation -> setConversation(a, c) }
            }
            .observeOn(mUiScheduler)
            .subscribe({}) { e: Throwable ->
                Log.e(TAG, "Error loading conversation", e)
                view?.goToHome()
            })
        view?.setReadIndicatorStatus(showReadIndicator())
    }

    override fun unbindView() {
        super.unbindView()
        mConversation = null
        mConversationUri = null
        mConversationDisposable?.let { conversationDisposable ->
            conversationDisposable.dispose()
            mConversationDisposable = null
        }
    }

    private fun setConversation(account: Account, conversation: Conversation) {
        Log.w(TAG, "setConversation " + conversation.aggregateHistory.size)
        if (mConversation == conversation) return
        mConversation = conversation
        mConversationSubject.onNext(conversation)
        val view = view
        view?.let { initView(account, conversation, it) }
    }

    fun pause() {
        mVisibilityDisposable.clear()
        mConversation?.isVisible = false
    }

    fun resume(isBubble: Boolean) {
        Log.w(TAG, "resume $mConversationUri")
        mVisibilityDisposable.clear()
        mVisibilityDisposable.add(mConversationSubject
            .subscribe({ conversation: Conversation ->
                conversation.isVisible = true
                updateOngoingCallView(conversation)
                mAccountService.getAccount(conversation.accountId)?.let { account ->
                    mConversationFacade.readMessages(account, conversation, !isBubble)}
            }) { e -> Log.e(TAG, "Error loading conversation", e) })
    }

    private fun initContact(account: Account, conversation: Conversation, mode: Conversation.Mode, view: ConversationView) {
        if (account.isJami) {
            Log.w(TAG, "initContact " + conversation.uri)
            if (mode === Conversation.Mode.Syncing) {
                view.switchToSyncingView()
            } else if (conversation.isSwarm || account.isContact(conversation)) {
                //if (conversation.isEnded())
                //    conversation.s
                view.switchToConversationView()
            } else {
                val uri = conversation.uri
                val req = account.getRequest(uri)
                if (req == null) {
                    view.switchToUnknownView(uri.rawUriString)
                } else {
                    view.switchToIncomingTrustRequestView(req.displayname)
                }
            }
        } else {
            view.switchToConversationView()
        }
        view.displayContact(conversation)
    }

    private fun initView(account: Account, c: Conversation, view: ConversationView) {
        Log.w(TAG, "initView " + c.uri + " " + c.mode)
        val disposable = mConversationDisposable?.apply { clear() } ?: CompositeDisposable().apply {
            mConversationDisposable = this
            mCompositeDisposable.add(this)
        }

        view.hideNumberSpinner()
        disposable.add(c.mode
            .switchMapSingle { mode: Conversation.Mode ->
                mContactService.getLoadedContact(c.accountId, c.contacts, true)
                    .observeOn(mUiScheduler)
                    .doOnSuccess { initContact(account, c, mode, view) }
            }
            .subscribe())
        disposable.add(c.mode
            .switchMap { mode: Conversation.Mode -> if (mode === Conversation.Mode.Legacy || mode === Conversation.Mode.OneToOne) c.contact!!.conversationUri else Observable.empty() }
            .observeOn(mUiScheduler)
            .subscribe { uri: Uri -> init(uri, account.accountID) })
        disposable.add(Observable.combineLatest(mHardwareService.connectivityState, mAccountService.getObservableAccount(account),
            { isConnected: Boolean, a: Account -> isConnected || a.isRegistered })
            .observeOn(mUiScheduler)
            .subscribe { isOk: Boolean ->
                this.view?.let { v ->
                    if (!isOk) v.displayNetworkErrorPanel() else if (!account.isEnabled) {
                        v.displayAccountOfflineErrorPanel()
                    } else {
                        v.hideErrorPanel()
                    }
                }
            })
        disposable.add(c.sortedHistory
            .observeOn(mUiScheduler)
            .subscribe({ conversation: List<Interaction> -> this.view?.refreshView(conversation) }) { e: Throwable ->
                Log.e(TAG, "Can't update element", e)
            })
        disposable.add(c.cleared
            .observeOn(mUiScheduler)
            .subscribe({ conversation: List<Interaction> -> this.view?.refreshView(conversation) }) { e: Throwable ->
                Log.e(TAG, "Can't update elements", e)
            })
        disposable.add(c.contactUpdates
            .switchMap { contacts: List<Contact> ->
                Observable.merge(mContactService.observeLoadedContact(c.accountId, contacts, true))
            }
            .observeOn(mUiScheduler)
            .subscribe { contact: Contact -> this.view?.updateContact(contact) })
        disposable.add(c.updatedElements
            .observeOn(mUiScheduler)
            .subscribe({ elementTuple ->
                val v = this.view ?: return@subscribe
                when (elementTuple.second) {
                    ElementStatus.ADD -> v.addElement(elementTuple.first)
                    ElementStatus.UPDATE -> v.updateElement(elementTuple.first)
                    ElementStatus.REMOVE -> v.removeElement(elementTuple.first)
                }
            }, { e: Throwable -> Log.e(TAG, "Can't update element", e) })
        )
        if (showTypingIndicator()) {
            disposable.add(c.composingStatus
                .observeOn(mUiScheduler)
                .subscribe { composingStatus: ComposingStatus -> this.view?.setComposingStatus(composingStatus) })
        }
        disposable.add(c.getLastDisplayed()
            .observeOn(mUiScheduler)
            .subscribe { interaction: Interaction -> this.view?.setLastDisplayed(interaction) })
        disposable.add(c.calls
            .observeOn(mUiScheduler)
            .subscribe({ updateOngoingCallView(c) }) { e: Throwable ->
                Log.e(TAG, "Can't update call view", e)
            })
        disposable.add(c.getColor()
            .observeOn(mUiScheduler)
            .subscribe({ integer: Int -> view.setConversationColor(integer) }) { e: Throwable ->
                Log.e(TAG, "Can't update conversation color", e)
            })
        disposable.add(c.getSymbol()
            .observeOn(mUiScheduler)
            .subscribe({ symbol: CharSequence -> view.setConversationSymbol(symbol) }) { e: Throwable ->
                Log.e(TAG, "Can't update conversation color", e)
            })
        disposable.add(account
            .getLocationUpdates(c.uri)
            .observeOn(mUiScheduler)
            .subscribe {
                Log.e(TAG, "getLocationUpdates: update")
                view?.showMap(c.accountId, c.uri.uri, false)
            }
        )
    }

    fun loadMore() {
        mConversationDisposable?.add(mAccountService.loadMore(mConversation!!).subscribe({}) {})
    }

    fun openContact() {
        mConversation?.let { conversation -> view?.goToContactActivity(conversation.accountId, conversation.uri) }
    }

    fun sendTextMessage(message: String?) {
        val conversation = mConversation
        if (message == null || message.isEmpty() || conversation == null) {
            return
        }
        val conference = conversation.currentCall
        if (conversation.isSwarm || conference == null || !conference.isOnGoing) {
            mConversationFacade.sendTextMessage(conversation, conversation.uri, message).subscribe()
        } else {
            mConversationFacade.sendTextMessage(conversation, conference, message)
        }
    }

    fun selectFile() {
        view!!.openFilePicker()
    }

    fun sendFile(file: File) {
        mCompositeDisposable.add(mConversationSubject.firstElement().subscribe({ conversation ->
            mConversationFacade.sendFile(conversation, conversation.uri, file).subscribe()
        }) {e -> Log.e(TAG, "Can't send file", e)})
    }

    /**
     * Gets the absolute path of the file dataTransfer and sends both the DataTransfer and the
     * found path to the ConversationView in order to start saving the file
     *
     * @param interaction an interaction representing a datat transfer
     */
    fun saveFile(interaction: Interaction) {
        val transfer = interaction as DataTransfer
        val fileAbsolutePath = deviceRuntimeService.getConversationPath(transfer).absolutePath
        view?.startSaveFile(transfer, fileAbsolutePath)
    }

    fun shareFile(interaction: Interaction) {
        val file = interaction as DataTransfer
        val path = deviceRuntimeService.getConversationPath(file)
        view?.shareFile(path, file.displayName)
    }

    fun openFile(interaction: Interaction) {
        val file = interaction as DataTransfer
        val path = deviceRuntimeService.getConversationPath(file)
        view?.openFile(path, file.displayName)
    }

    fun acceptFile(transfer: DataTransfer) {
        view?.acceptFile(mConversation!!.accountId, mConversationUri!!, transfer)
    }

    fun refuseFile(transfer: DataTransfer) {
        view!!.refuseFile(mConversation!!.accountId, mConversationUri!!, transfer)
    }

    fun deleteConversationItem(element: Interaction) {
        mConversationFacade.deleteConversationItem(mConversation!!, element)
    }

    fun cancelMessage(message: Interaction) {
        mConversationFacade.cancelMessage(message)
    }

    private fun sendTrustRequest() {
        val conversation = mConversation ?: return
        val contact = conversation.contact ?: return
        contact.status = Contact.Status.REQUEST_SENT
        mVCardService.loadSmallVCardWithDefault(conversation.accountId, VCardService.MAX_SIZE_REQUEST)
            .subscribeOn(Schedulers.computation())
            .subscribe({ vCard -> mAccountService.sendTrustRequest(conversation, contact.uri, Blob.fromString(VCardUtils.vcardToString(vCard)))})
            { mAccountService.sendTrustRequest(conversation, contact.uri, null) }
    }

    fun clickOnGoingPane() {
        val conf = mConversation?.currentCall
        if (conf != null) {
            view?.goToCallActivity(conf.id)
        } else {
            view?.displayOnGoingCallPane(false)
        }
    }

    fun goToCall(audioOnly: Boolean) {
        if (audioOnly && !mHardwareService.hasMicrophone()) {
            view!!.displayErrorToast(Error.NO_MICROPHONE)
            return
        }
        mCompositeDisposable.add(mConversationSubject
            .firstElement()
            .subscribe { conversation: Conversation ->
                val view = view
                if (view != null) {
                    val conf = conversation.currentCall
                    if (conf != null && conf.participants.isNotEmpty()
                        && conf.participants[0].callStatus !== Call.CallStatus.INACTIVE
                        && conf.participants[0].callStatus !== Call.CallStatus.FAILURE) {
                        view.goToCallActivity(conf.id)
                    } else {
                        view.goToCallActivityWithResult(conversation.accountId, conversation.uri, conversation.contact!!.uri, audioOnly)
                    }
                }
            })
    }

    private fun updateOngoingCallView(conversation: Conversation?) {
        val conf = conversation?.currentCall
        view?.displayOnGoingCallPane(conf != null && (conf.state === Call.CallStatus.CURRENT || conf.state === Call.CallStatus.HOLD || conf.state === Call.CallStatus.RINGING))
    }

    fun onBlockIncomingContactRequest() {
        mConversation?.let { conversation ->
            mConversationFacade.discardRequest(conversation.accountId, conversation.uri)
            mAccountService.removeContact(conversation.accountId, conversation.uri.host, true)
        }
        view?.goToHome()
    }

    fun onRefuseIncomingContactRequest() {
        mConversation?.let { conversation ->
            mConversationFacade.discardRequest(conversation.accountId, conversation.uri)
        }
        view?.goToHome()
    }

    fun onAcceptIncomingContactRequest() {
        mConversation?.let { conversation ->
            mConversationFacade.acceptRequest(conversation.accountId, conversation.uri)
        }
        view?.switchToConversationView()
    }

    fun onAddContact() {
        sendTrustRequest()
        view?.switchToConversationView()
    }

    fun noSpaceLeft() {
        Log.e(TAG, "configureForFileInfoTextMessage: no space left on device")
        view?.displayErrorToast(Error.NO_SPACE_LEFT)
    }

    fun setConversationColor(color: Int) {
        mCompositeDisposable.add(mConversationSubject
            .firstElement()
            .subscribe { conversation: Conversation -> conversation.setColor(color) })
    }

    fun setConversationSymbol(symbol: CharSequence) {
        mCompositeDisposable.add(mConversationSubject.firstElement()
            .subscribe { conversation -> conversation.setSymbol(symbol) })
    }

    fun cameraPermissionChanged(isGranted: Boolean) {
        if (isGranted && mHardwareService.isVideoAvailable) {
            mHardwareService.initVideo()
                .onErrorComplete()
                .subscribe()
        }
    }

    fun shareLocation() {
        mCompositeDisposable.add(mConversationSubject.firstElement()
            .subscribe { conversation -> view?.startShareLocation(conversation.accountId, conversation.uri.uri) })
    }

    fun showPluginListHandlers() {
        view?.showPluginListHandlers(mConversation!!.accountId, mConversationUri!!.uri)
    }

    val path: Pair<String, String>
        get() = Pair(mConversation!!.accountId, mConversationUri!!.uri)

    fun onComposingChanged(hasMessage: Boolean) {
        if (showTypingIndicator()) {
            mConversation?.let { conversation ->
                mConversationFacade.setIsComposing(conversation.accountId, conversation.uri, hasMessage)
            }
        }
    }

    private fun showTypingIndicator(): Boolean {
        return mPreferencesService.settings.isAllowTypingIndicator
    }

    private fun showReadIndicator(): Boolean {
        return mPreferencesService.settings.isAllowReadIndicator
    }

    companion object {
        private val TAG = ConversationPresenter::class.simpleName!!
    }

    init {
        mCompositeDisposable.add(mVisibilityDisposable)
    }
}